certbot/plugins/dns_cloudflare.py

```diff
@@ -0,0 +1,323 @@
+
+        if self.email:
+            setattr(self.config, self.dest('email'), self.email)
+        else:
+            raise errors.PluginError('Cloudflare account email address required to proceed.')
+
+        configured_api_key = self.conf('api-key')
+        if configured_api_key:
+            self.api_key = configured_api_key
+        else:
+            self.api_key = self._prompt_for_data('API key')
+
+        if self.api_key:
+            setattr(self.config, self.dest('api-key'), self.api_key)
+        else:
+            raise errors.PluginError('Cloudflare account API key required to proceed.')
+
+    @staticmethod
+    def _prompt_for_data(label):
+        """
+        Prompt the user for a piece of information.
+
+        :param string label: The user-friendly label for this piece of information.
+        :returns: The user's response (guaranteed non-empty).
+        :rtype: string
+        """
+
+        display = zope.component.getUtility(interfaces.IDisplay)
+
+        while True:
+            code, response = display.input(
+                'Input Cloudflare account {0}'.format(label),
+                force_interactive=True)
+            if code == display_util.HELP:
+                # Displaying help is not currently implemented
+                return None
+            elif code == display_util.CANCEL:
+                return None
+            else:  # code == display_util.OK
+                if not response:
+                    display.notification('Please enter an {0}.'.format(label), pause=False)
+                else:
+                    return response
+
+    def _get_cloudflare_client(self):
+        """
+        Helper method to construct a `_CloudflareClient`.
+
+        Uses configured credentials.
+
+        :return: a new
+        :rtype: `_CloudflareClient`
+        """
+
+        return _CloudflareClient(self.email, self.api_key)
+
+
+class _CloudflareClient(object):
+    """
+    Encapsulates all communication with the Cloudflare API.
+    """
+
+    def __init__(self, email, api_key):
+        self.cf = CloudFlare.CloudFlare(email, api_key)
+
+    def add_txt_record(self, domain, record_name, record_content, record_ttl):
+        """
+        Add a TXT record using the supplied information.
+
+        :param string domain: The domain to use to look up the Cloudflare zone.
+        :param string record_name: The record name (typically beginning with '_acme-challenge.').
+        :param string record_content: The record content (typically the challenge validation).
+        :param int record_ttl: The record TTL (number of seconds that the record may be cached).
+        :raises: errors.PluginError if an error occurs communicating with the Cloudflare API
+        """
+
+        zone_id = self._find_zone_id(domain)
+
+        data = {'type': 'TXT',
+                'name': record_name,
+                'content': record_content,
+                'ttl': record_ttl}
```

Comment:
> I think changing the way Certbot is packaged/distributed so that dependencies aren't a huge headache may be a more valuable investment. What do you think of #4537 (which turns out to be a duplicate of #334) as a way to address dependency problems?

I'd love to do that but some of us have been pushing for this for a year+ at this point and haven't been able to get approval. There's been a lot of renewed interest and frustration recently so it's possible things will change this time, but I'm not super confident this will happen in the short term. I'm a little hesitant to block this PR on those decisions, but if you want I can bring them up in the next week or two and see what we can accomplish.

What I'd like to do is reach out the Lexicon package maintainer, explain our situation and what we'd like to do, and see how receptive they are to the changes we'd like to make. This can help inform what we do here. I think it'd be good if you talked to them since you're the most familiar with the library but I can reach out first if you prefer.

Also, while I understand the inefficiency of the guess-and-check method of determining a DNS zone to modify with Lexicon, what are the other downsides to this approach? Is the code significantly more complex or error prone? How does doing this negatively impact the user? I just want to make sure I properly understand the problems here so we can make the right decision.

I talked about this with Erica for a bit and I think we should have a design meeting about packaging these plugins this week. My inclination right now is to put this plugin as an external package if we can reasonably move it into core Certbot if we want later without causing problems for users. I think this is doable but I want to talk and think about it to make sure. Please let me know what you think!